src/main/java/com/example/onlinebookstore/service/user/UserServiceImpl.java

```diff
@@ -0,0 +1,46 @@
+package com.example.onlinebookstore.service.user;
+
+import com.example.onlinebookstore.dto.user.UserRegistrationRequestDto;
+import com.example.onlinebookstore.dto.user.UserResponseDto;
+import com.example.onlinebookstore.exception.RegistrationException;
+import com.example.onlinebookstore.mapper.UserMapper;
+import com.example.onlinebookstore.model.Role;
+import com.example.onlinebookstore.model.RoleName;
+import com.example.onlinebookstore.model.User;
+import com.example.onlinebookstore.repository.RoleRepository;
+import com.example.onlinebookstore.repository.UserRepository;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
+import lombok.RequiredArgsConstructor;
+import org.springframework.security.crypto.password.PasswordEncoder;
+import org.springframework.stereotype.Service;
+
+@Service
+@RequiredArgsConstructor
+public class UserServiceImpl implements UserService {
+    private final UserRepository userRepository;
+    private final PasswordEncoder passwordEncoder;
+    private final UserMapper userMapper;
+    private final RoleRepository roleRepository;
+
+    @Override
+    public UserResponseDto register(UserRegistrationRequestDto request)
+            throws RegistrationException {
+        if (userRepository.findByEmail(request.email()).isPresent()) {
+            throw new RegistrationException("Unable to complete registration");
```

Comment:
You don't need to retrieve full user by email. Instead of this you can define existByEmail method In userRepository that return boolean value

Reply:
I didn't quite understand the comment. While pouring without this correction